server/src/main/java/org/opensearch/index/shard/StoreRecovery.java

```diff
@@ -791,4 +795,26 @@ private void bootstrap(final IndexShard indexShard, final Store store) throws IO
         );
         store.associateIndexWithNewTranslog(translogUUID);
     }
+
+    /*
+    Blocks the calling thread,  waiting for the remote store to get synced till internal Remote Upload Timeout
+     */
+    private void waitForRemoteStoreSync(IndexShard indexShard) {
+        if (indexShard.shardRouting.primary() == false) {
+            return;
+        }
+        long startNanos = System.nanoTime();
+
+        while (System.nanoTime() - startNanos < indexShard.getRecoverySettings().internalRemoteUploadTimeout().nanos()) {
```

Comment:
What happens if the upload takes longer than the timeout? I am assuming we fail the recoveries. How do we ensure we cancel all in-flight data transfers and avoid wasted work resulting in concurrent uploads while recovery has been cancelled

Reply:
We don't cancel in-flight data transfers , as there is such communication to `RemoteStoreRefreshListener` available. But after an attempt of uploading it , it wouldn't again try to upload as the shard is no longer present . 

I think we might not need this mechanism immediately as the 1 hour timeout is quite generous to upload the data. If a user does see it, they have the mechanism to tune the timeout as well .